include/realtime_tools/realtime_barrier.hpp

```diff
@@ -0,0 +1,468 @@
+    }
+
+    inline void reset()
+    {
+      if (lock_strategy::owns_lock()) {
+        lock_strategy::unlock();
+      }
+      mem_ = nullptr;
+      obj_ = nullptr;
+    }
+
+    inline bool new_data_available() const
+    {
+      return mem_->new_data_available();
+    }
+
+    inline void new_data_available(bool avail)
+    {
+      if (!lock_strategy::owns_lock()) {
+        throw std::runtime_error("can't modify an unlocked MemoryBarrier");
+      }
+      mem_->new_data_available_ = avail;
+    }
+
+    inline void swap()
+    {
+      if (!std::unique_lock<std::mutex>::owns_lock()) {
+        throw std::runtime_error("can't swap an unlocked MemoryBarrier");
+      }
+      mem_->swap();
+      obj_ = nullptr;   // must explicitly set to null so _get doesnt return cached value
+      obj_ = _get();
+    }
+
+    inline explicit operator bool() const
+    {return mem_ != nullptr && get();}
+
+    inline bool operator==(std::nullptr_t) const
+    {return mem_ == nullptr && get();}
+
+    inline bool operator!=(std::nullptr_t) const
+    {return mem_ != nullptr && get();}
+
+    // smartptr semantics
+    inline T * get()
+    {return _get();}
+
+    inline const T * get() const
+    {return _get();}
+
+    inline T * operator->()
+    {return get();}
+
+    inline const T * operator->() const
+    {return get();}
+
+    inline T & operator*()
+    {return *get();}
+
+    inline const T & operator*() const
+    {return *get();}
+
+private:
+    MemoryBarrierType * mem_;
+    mutable T * obj_;
+
+    template<class realtimeness = access_mode>
+    inline typename std::enable_if_t<std::is_trivially_assignable<realtime, realtimeness>::value,
+      T *>
+    _get()
+    {
+      return obj_ ?
+             obj_ :
+             lock_strategy::owns_lock() ?
+             obj_ = mem_->rt_ :
+               nullptr;
+    }
+
+    template<class realtimeness = access_mode>
+    inline typename std::enable_if_t<std::is_trivially_assignable<non_realtime,
+      realtimeness>::value, T *>
+    _get()
+    {
+      return obj_ ?
+             obj_ :
+             lock_strategy::owns_lock() ?
+             obj_ = mem_->nrt_ :
+               nullptr;
+    }
+
+    template<class realtimeness = access_mode>
+    inline typename std::enable_if_t<std::is_trivially_assignable<realtime, realtimeness>::value,
+      const T *>
+    _get() const
+    {
+      return obj_ ?
+             obj_ :
+             lock_strategy::owns_lock() ?
+             obj_ = mem_->rt_ :
+               nullptr;
+    }
+
+    template<class realtimeness = access_mode>
+    inline typename std::enable_if_t<std::is_trivially_assignable<non_realtime,
+      realtimeness>::value, const T *>
+    _get() const
+    {
+      return obj_ ?
+             obj_ :
+             lock_strategy::owns_lock() ?
+             obj_ = mem_->nrt_ :
+               nullptr;
+    }
+  };
+
+
+  MemoryBarrier()
+  : nrt_(new T()), rt_(new T()), polarity_(false), new_data_available_(false)
+  {
+  }
```

Comment:
ok, I'm trying to make it through all this but at this point I'm just kind of tired and frustrated and it's midnight. I'm having trouble separating the actual legitimate issues with all the stylistic ones. I really do appreciate the attention to detail that discovered at least 2 bugs but It would have been far easier to just commit a new file with all your recommended changes that I could review. There are style changes for almost every inch of the file. We have a linter and styler that enforces Ros2 policies. I don't like them all either but it's there for a reason.